#Homework assignment #3

These problem sets focus on using the Beautiful Soup library to scrape web pages.

##Problem Set #1: Basic scraping

I've made a web page for you to scrape. It's available [here](http://static.decontextualize.com/widgets2015.html). The page concerns the catalog of a famous [widget](http://en.wikipedia.org/wiki/Widget) company. You'll be answering several questions about this web page. First off, in the cell below, write some code so that you end up with a variable called `html_str` that contains the HTML source code of the page.

In [2]:
import urllib
html_str = urllib.urlopen("http://static.decontextualize.com/widgets2015.html").read()
html_str

'<!doctype html>\n<html>\n\t<head>\n\t\t<title>Fearless Widget Factory Catalog</title>\n\t\t<style type="text/css">\n\t\t\ttable.widgetlist {\n\t\t\t\tborder-collapse: collapse;\n\t\t\t\tborder: 1px #333 solid;\n\t\t\t\twidth: 50%;\n\t\t\t}\n\t\t\ttable.widgetlist td {\n\t\t\t\tborder: 1px #333 solid;\n\t\t\t\tpadding: 5px;\n\t\t\t}\n\t\t\ttable.widgetlist thead {\n\t\t\t\tbackground: #333;\n\t\t\t\tcolor: white;\n\t\t\t}\n\t\t</style>\n\t</head>\n\t<body>\n\n<h1>Fearless Widget Factory</h1>\n\n<p><i>Purveyor of fearless widgets since 1956</i></p>\n\n<h2>Widget Catalog</h2>\n\n<p>Please call <span class="tel">212-555-9912</span> for latest prices.</p>\n\n<h3>Forensic Widgets</h3>\n\n<table class="widgetlist">\n\t<thead>\n\t\t<td>Part No.</td>\n\t\t<td>Name</td>\n\t\t<td>Price</td>\n\t\t<td>Quantity in warehouse</td>\n\t</thead>\n  <tr class="winfo">\n    <td class="partno">C1-9476</td>\n    <td class="wname"><b>Skinner Widget</b></td>\n    <td class="price">$2.70</td>\n    <td class="q

Now, in the cell below, use Beautiful Soup to write an expression that evaluates to the number of `<h3>` tags contained in `widgets.html`. I've added the Beautiful Soup import statement for you.

In [3]:
from bs4 import BeautifulSoup
document = BeautifulSoup(html_str)
h3_tag = document.find_all('h3')
len(h3_tag)
[tag.string for tag in h3_tag]

[u'Forensic Widgets', u'Mood widgets', u'Hallowed widgets']

Now, in the cell below, write an expression or series of statements that displays the telephone number beneath the "Widget Catalog" header.

In [4]:
#<span class="tel">212-555-9912</span> 
checkup_tags = document.find_all('span', attrs={'class': 'tel'})
[tag.string for tag in checkup_tags]


[u'212-555-9912']

In the cell below, use Beautiful Soup to write some code that creates a list of the names of all the widgets on the page. After your code has executed, `widget_names` should evaluate to a list that looks like this (though not necessarily in this order):

    [u'Skinner Widget',
     u'Widget For Furtiveness',
     u'Widget For Strawman',
     u'Manicurist Widget',
     u'Infinite Widget',
     u'Yellow-Tipped Widget',
     u'Unshakable Widget',
     u'Self-Knowledge Widget',
     u'Widget For Cinema']

In [5]:
widget_tags = document.find_all('td', attrs={'class': 'wname'})
[widget_tag.string for widget_tag in widget_tags]


[u'Skinner Widget',
 u'Widget For Furtiveness',
 u'Widget For Strawman',
 u'Manicurist Widget',
 u'Infinite Widget',
 u'Yellow-Tipped Widget',
 u'Unshakable Widget',
 u'Self-Knowledge Widget',
 u'Widget For Cinema']

##Problem set #2: Widget dictionaries

For this problem set, we'll continue to use the HTML page from the previous problem set. In the cell below, I've made an empty list and assigned it to a variable called `widgets`. Write code that populates this list with dictionaries, one dictionary per widget in the source file. The keys of each dictionary should be `partno`, `widgetname`, `price`, and `quantity`, and the value for each of the keys should be the value for the corresponding column for each row. After executing the cell, your list should look something like this:

    [{'partno': u'C1-9476',
      'price': u'$2.70',
      'quantity': u'512',
      'wname': u'Skinner Widget'},
     {'partno': u'JDJ-32/V',
      'price': u'$9.36',
      'quantity': u'967',
      'wname': u'Widget For Furtiveness'},
      ...several items omitted...
     {'partno': u'5B-941/F',
      'price': u'$13.26',
      'quantity': u'919',
      'wname': u'Widget For Cinema'}]

And this expression:

    widgets[5]['partno']
    
... should evaluate to:

    u'MZ-556/B'
    
(Hint: This could get a little messy! Each column in the table might need to be treated a bit differently to get the appropriate strings out.)

In [6]:
widgetlist = []
for widget in document.find_all('tr'):
    widget_dict = {}
    
    name_widget = widget.find('td', attrs={'class': 'wname'}).string
    widget_dict['name'] = name_widget
    
    partno_widget = widget.find('td', attrs={'class': 'partno'}).string
    widget_dict['partno'] = partno_widget
    
    price_widget = widget.find('td', attrs={'class': 'price'}).string
    widget_dict['price'] = price_widget
    
    quantity_widget = widget.find('td', attrs={'class': 'quantity'}).string
    widget_dict['quantity'] = quantity_widget
    
    widgetlist.append(widget_dict)
widgetlist


[{'name': u'Skinner Widget',
  'partno': u'C1-9476',
  'price': u'$2.70',
  'quantity': u'512'},
 {'name': u'Widget For Furtiveness',
  'partno': u'JDJ-32/V',
  'price': u'$9.36',
  'quantity': u'967'},
 {'name': u'Widget For Strawman',
  'partno': u'YP4-325/J',
  'price': u'$5.17',
  'quantity': u'787'},
 {'name': u'Manicurist Widget',
  'partno': u'VK-486',
  'price': u'$8.97',
  'quantity': u'441'},
 {'name': u'Infinite Widget',
  'partno': u'R4K-990',
  'price': u'$11.73',
  'quantity': u'320'},
 {'name': u'Yellow-Tipped Widget',
  'partno': u'MZ-556/B',
  'price': u'$2.35',
  'quantity': u'948'},
 {'name': u'Unshakable Widget',
  'partno': u'QV-730',
  'price': u'$3.76',
  'quantity': u'59'},
 {'name': u'Self-Knowledge Widget',
  'partno': u'T1-9731',
  'price': u'$7.11',
  'quantity': u'790'},
 {'name': u'Widget For Cinema',
  'partno': u'5B-941/F',
  'price': u'$13.26',
  'quantity': u'919'}]

In [7]:
widgetlist[5]['partno']

u'MZ-556/B'

In the cell below, duplicate your code from the previous question. Modify the code to ensure that the values for `price` and `quantity` in each dictionary are floating-point numbers and integers, respectively. I.e., after executing the cell, your code should display something like this:

    [{'partno': 'C1-9476',
      'price': 2.7,
      'quantity': 512,
      'widgetname': 'Skinner Widget'},
     {'partno': 'JDJ-32/V',
      'price': 9.36,
      'quantity': 967,
      'widgetname': 'Widget For Furtiveness'},
     ... some items omitted ...
     {'partno': '5B-941/F',
      'price': 13.26,
      'quantity': 919,
      'widgetname': 'Widget For Cinema'}]

(Hint: Use the `float()` and `int()` functions. You may need to use string slices to convert the `price` field to a floating-point number.)

In [27]:

html_str = urllib.urlopen("http://static.decontextualize.com/widgets2015.html").read()
document = BeautifulSoup(html_str)

widgetlist2 = []
for widget in document.find_all('tr'):
    widget_dict = {}
    
    name_widget = widget.find('td', attrs={'class': 'wname'}).string
    widget_dict['name'] = name_widget
    
    partno_widget = widget.find('td', attrs={'class': 'partno'}).string
    widget_dict['partno'] = partno_widget
    
    price_widget = widget.find('td', attrs={'class': 'price'}).string
    price_widget_num = float(price_widget[1:])
    widget_dict['price'] = price_widget_num
    
    quantity_widget = widget.find('td', attrs={'class': 'quantity'}).string
    widget_dict['quantity'] = int(quantity_widget)
    
    widgetlist2.append(widget_dict)
print widgetlist2


[{'partno': u'C1-9476', 'price': 2.7, 'name': u'Skinner Widget', 'quantity': 512}, {'partno': u'JDJ-32/V', 'price': 9.36, 'name': u'Widget For Furtiveness', 'quantity': 967}, {'partno': u'YP4-325/J', 'price': 5.17, 'name': u'Widget For Strawman', 'quantity': 787}, {'partno': u'VK-486', 'price': 8.97, 'name': u'Manicurist Widget', 'quantity': 441}, {'partno': u'R4K-990', 'price': 11.73, 'name': u'Infinite Widget', 'quantity': 320}, {'partno': u'MZ-556/B', 'price': 2.35, 'name': u'Yellow-Tipped Widget', 'quantity': 948}, {'partno': u'QV-730', 'price': 3.76, 'name': u'Unshakable Widget', 'quantity': 59}, {'partno': u'T1-9731', 'price': 7.11, 'name': u'Self-Knowledge Widget', 'quantity': 790}, {'partno': u'5B-941/F', 'price': 13.26, 'name': u'Widget For Cinema', 'quantity': 919}]


Great! I hope you're having fun. In the cell below, write an expression that uses the `widgets` list created in the cell above to calculate the total number of widgets that the factory has in its warehouse.

Expected output: `5743`

In [20]:
quantitylist = []
for i in range(9):
    numbs = (widgetlist2[i]['quantity'])
    quantitylist.append(numbs)
sum(quantitylist)


5743

In the cell below, write an expression (or series of expressions) that evaluates to a list the names of widgets containing only those widgets whose price is above $9.30.

Expected output: `[u'Widget For Furtiveness', u'Infinite Widget', u'Widget For Cinema']`

In [287]:
morelist = []
for i in range(9):
    if widgetlist2[i]['price'] > 9.3: 
        morelist.append(widgetlist2[i]['name'])
print morelist

[u'Widget For Furtiveness', u'Infinite Widget', u'Widget For Cinema']


##Problem set #3: Sibling rivalries

In the following problem set, you will yet again be working with the data in `widgets.html`. In order to accomplish the tasks in this problem set, you'll need to learn about Beautiful Soup's `.find_next_sibling()` method. Here's some information about that method, cribbed from the notes:

Often, the tags we're looking for don't have a distinguishing characteristic, like a class attribute, that allows us to find them using `.find()` and `.find_all()`, and the tags also aren't in a parent-child relationship. This can be tricky! For example, take the following HTML snippet, (which I've assigned to a string called `example_html`):

If our task was to create a dictionary that maps the name of the cheese to the description that follows in the `<p>` tag directly afterward, we'd be out of luck. Fortunately, Beautiful Soup has a `.find_next_sibling()` method, which allows us to search for the next tag that is a sibling of the tag you're calling it on (i.e., the two tags share a parent), that also matches particular criteria. So, for example, to accomplish the task outlined above:

In [74]:
# example_doc = BeautifulSoup(example_html)
# cheese_dict = {}
# for h2_tag in example_doc.find_all('h2'):
#     cheese_name = h2_tag.string
#     cheese_desc_tag = h2_tag.find_next_sibling('p')
#     cheese_dict[cheese_name] = cheese_desc_tag.string

# cheese_name

With that knowledge in mind, let's go back to our widgets. In the cell below, write code that uses Beautiful Soup, and in particular the `.find_next_sibling()` method, to find out how many widgets are in the table *just beneath* the header "Hallowed Widgets." (You can tell by looking at the page that there are four such widgets. But this is a programming class, so we have to write a program to do it.)

In [285]:
import urllib
html_str = urllib.urlopen("http://static.decontextualize.com/widgets2015.html").read()
document = BeautifulSoup(html_str)
widget = []

category_counts = {}
for h3_tag in document.find_all('h3'):
    h3_string = h3_tag.string
    if h3_string == 'Hallowed Widgets':
        tabletag = h3_tag.find_next_sibling('table')
        trtags = tabletag.find_all('tr')
        count = len(trtags)
count

4

Okay, now, the final task. If you can accomplish this, you are truly an expert web scraper. I'll have little web scraper certificates made up and I'll give you one, if you manage to do this thing. And I know you can do it!

In the cell below, I've created a variable `category_counts` and assigned to it an empty dictionary. Write code to populate this dictionary so that its keys are "categories" of widgets (e.g., the contents of the `<h3>` tags on the page: "Forensic Widgets", "Mood widgets", "Hallowed Widgets") and the value for each key is the number of widgets that occur in that category. I.e., after your code has been executed, the dictionary `category_counts` should look like this:

    {u'Forensic Widgets': 3, u'Hallowed widgets': 4, u'Mood widgets': 2}


In [282]:
import urllib
html_str = urllib.urlopen("http://static.decontextualize.com/widgets2015.html").read()
document = BeautifulSoup(html_str)

category_counts = {}
for h3_tag in document.find_all('h3'):
    h3_string = h3_tag.string
    tabletag = h3_tag.find_next_sibling('table')
    trtags = tabletag.find_all('tr')
    count = len(trtags)
    category_counts[h3_string] = count
category_counts

    

{u'Forensic Widgets': 3, u'Hallowed widgets': 4, u'Mood widgets': 2}

Congratulations! You're done.